### PR TITLE
Supporting EIPs with specified BYOIP addresses

### LINKED ISSUE #8004: Specify explicit address from BYOIP pool
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

We have a CIDR allocated to us by ARIN. We used the [BYOIP](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html) functionality to bring the addresses to AWS. Now we want to use the addresses, specifying a particular IP from our pool. Currently one can only use a random address within the pool, not a specific address.

Looking at the [AWS API documentation](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AllocateAddress.html) one needs to simply include the `Address` parameter instead of `PublicIpv4Pool` and AWS will assign the IP from the proper pool.

### New or Affected Resource(s)

* aws_eip

### Potential Terraform Configuration

```hcl
resource "aws_eip" "byoip-ip" {
  vpc              = true
  address          = "1.2.3.4"
}
```

### References

* #6251
* #6521
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AllocateAddress.html

### PR DESCRIPTION
**Pull request by arwilczek90**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #8004

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
ENHANCEMENTS

* Added support for EIPs with a specified BYOIP address
```

Output from acceptance testing:

```
$  make testacc TEST=./aws TESTARGS='-run=TestAccAWSEIP_BYOIPAddress'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSEIP_BYOIPAddress -timeout 120m
=== RUN   TestAccAWSEIP_BYOIPAddress_default
=== PAUSE TestAccAWSEIP_BYOIPAddress_default
=== RUN   TestAccAWSEIP_BYOIPAddress_custom
--- PASS: TestAccAWSEIP_BYOIPAddress_custom (23.06s)
=== RUN   TestAccAWSEIP_BYOIPAddress_custom_with_PublicIpv4Pool
--- PASS: TestAccAWSEIP_BYOIPAddress_custom_with_PublicIpv4Pool (21.54s)
=== CONT  TestAccAWSEIP_BYOIPAddress_default
--- PASS: TestAccAWSEIP_BYOIPAddress_default (22.48s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       67.099s

```

### CODE REVIEW

**Comment by arwilczek90** [comment]
This has been tested with my company's IP block.

**Comment by ravenbyron** [comment]
I have tested this and it makes my workflow 10000% easier!!! Thanks

**Comment by ravenbyron** [comment]
Any idea when this may get reviewed/merged?

**Comment by arwilczek90** [comment]
any idea when we can see this reviewed?

**Comment by hsmithatemma** [comment]
This is a great idea!

**Comment by stobias123** [comment]
Bump - really like this feature

**Comment by z0ph** [comment]
This will be really useful for us. Thanks

**Comment by robzr** [comment]
Please consider adding this!

**Comment by aaronmell** [comment]
Why hasn't this been merged?

**Comment by arwilczek90** [comment]
I'm waiting on code review, I don't know why it hasn't gotten any review though.

**Comment by aaronmell** [comment]
@arwilczek90 Sorry, that wasn't directed at you. I've faced a similar issue with a PR in a different provider and I was venting a little bit in that regard.

**Comment by arwilczek90** [comment]
@aaronmell No worries 😃

**Comment by mnovacyu** [comment]
Would greatly appreciate having this! As an alternative, we've had to create a null_resource which calls a custom python script. When can this be reviewed?

**Comment by ravenbyron** [comment]
Can we please get an update on this !!!!

**Comment by gyoza** [comment]
Any chance that this will ever get added?

**Comment by hsmithatemma** [comment]
Any updates on this front? We just tripped over the 1year mark.

**Comment by ravenbyron** [comment]
@bflad any chance we can get a review and merge on this. If you need to test I have access to a dev env that has a set of BYOIPs that I could share.  (I am pinging you as you merged a similar pr for outpost)

**Comment by midN** [comment]
I guess it's time to migrate to CDK?
Terraform is now reaching that CloudFormation threshold with features lagging behing over a year

**Comment by ravenbyron** [comment]
Any chance for any feed back on this? cc: @bflad

**Comment by gwvandesteeg** [comment]
This looks good, code diff looks good, just needs an appropriate person to merge, be nice if it was done before my BYOIP block goes live, manually creating the EIPs without terraform is going to suck in managing it.

**Comment by midN** [comment]
Can we get some movement on this?
It's been open for almost 2 years now, is that the new state of Terraform where updates can't be merged in for 2 years despite having an open PR?

@YakDriver @bflad

**Comment by YakDriver** [comment]
One concern with this PR is the ability to perform meaningful testing of this new functionality in an automated way. Currently we do not have access to test with a public block of IPs. Perhaps the risk/reward still weighs in favor of adding the functionality. Do you have any suggestions?

Potential ideas for ways forward:
1. Some workaround for a BYOIP (using another EIP?)
2. Rather than nightly, automated tests, rely on occasional manual tests
3. Procure a public block for testing

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-byoip.html#byoip-provision

**Comment by bflad** [comment]
In the past when we have had acceptance tests which the HashiCorp maintainers cannot reasonably run, we have provided code reviews/updates for consistency, then asked the contributor and another community member to run and verify the acceptance testing. Might be the best path forward here.

**Comment by arwilczek90** [comment]
Unfortunately I'm not longer at the company I wrote this for(and like most people don't have an IPv4 block to play with), but @ravenbyron or @hsmithatemma are still there and possibly have access to a couple of IP addresses that can be used to provide test results.

**Review by YakDriver — APPROVED** [review]
Looks great! 🎉 

Output of acceptance tests (GovCloud):
```
--- PASS: TestAccAWSEIP_basic (32.33s)
--- PASS: TestAccAWSEIP_BYOIPAddress_default (32.32s)
--- PASS: TestAccAWSEIP_disappears (25.79s)
--- PASS: TestAccAWSEIP_Instance (137.48s)
--- PASS: TestAccAWSEIP_Instance_associatedUserPrivateIP (236.63s)
--- PASS: TestAccAWSEIP_Instance_notAssociated (146.57s)
--- PASS: TestAccAWSEIP_Instance_reassociate (122.72s)
--- PASS: TestAccAWSEIP_networkBorderGroup (32.43s)
--- PASS: TestAccAWSEIP_NetworkInterface (85.66s)
--- PASS: TestAccAWSEIP_NetworkInterface_twoEIPsOneInterface (85.66s)
--- PASS: TestAccAWSEIP_PublicIPv4Pool_default (32.33s)
--- PASS: TestAccAWSEIP_Tags_EC2VPC_withoutVPCTrue (45.92s)
--- PASS: TestAccAWSEIP_Tags_EC2VPC_withVPCTrue (45.99s)
--- PASS: TestAccAWSEIPAssociation_basic (164.12s)
--- PASS: TestAccAWSEIPAssociation_disappears (100.33s)
--- PASS: TestAccAWSEIPAssociation_instance (126.54s)
--- PASS: TestAccAWSEIPAssociation_networkInterface (80.73s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (92.08s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom (0.00s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom_with_PublicIpv4Pool (0.00s)
--- SKIP: TestAccAWSEIP_carrierIP (1.66s)
--- SKIP: TestAccAWSEIP_CustomerOwnedIPv4Pool (2.05s)
--- SKIP: TestAccAWSEIP_Instance_ec2Classic (2.73s)
--- SKIP: TestAccAWSEIP_PublicIPv4Pool_custom (0.00s)
--- SKIP: TestAccAWSEIP_Tags_EC2Classic_withoutVPCTrue (2.71s)
--- SKIP: TestAccAWSEIP_Tags_EC2Classic_withVPCTrue (2.73s)
--- SKIP: TestAccAWSEIPAssociation_ec2Classic (0.00s)
```

Output of acceptance tests (`us-west-2`):
```
--- PASS: TestAccAWSEIP_basic (33.09s)
--- PASS: TestAccAWSEIP_BYOIPAddress_default (32.61s)
--- PASS: TestAccAWSEIP_carrierIP (33.82s)
--- PASS: TestAccAWSEIP_disappears (26.80s)
--- PASS: TestAccAWSEIP_Instance (131.53s)
--- PASS: TestAccAWSEIP_Instance_associatedUserPrivateIP (240.46s)
--- PASS: TestAccAWSEIP_Instance_ec2Classic (238.25s)
--- PASS: TestAccAWSEIP_Instance_notAssociated (168.78s)
--- PASS: TestAccAWSEIP_Instance_reassociate (141.37s)
--- PASS: TestAccAWSEIP_networkBorderGroup (17.84s)
--- PASS: TestAccAWSEIP_NetworkInterface (84.06s)
--- PASS: TestAccAWSEIP_NetworkInterface_twoEIPsOneInterface (84.78s)
--- PASS: TestAccAWSEIP_PublicIPv4Pool_default (30.83s)
--- PASS: TestAccAWSEIP_Tags_EC2Classic_withoutVPCTrue (13.65s)
--- PASS: TestAccAWSEIP_Tags_EC2Classic_withVPCTrue (42.84s)
--- PASS: TestAccAWSEIP_Tags_EC2VPC_withoutVPCTrue (45.06s)
--- PASS: TestAccAWSEIP_Tags_EC2VPC_withVPCTrue (45.09s)
--- PASS: TestAccAWSEIPAssociation_basic (230.01s)
--- PASS: TestAccAWSEIPAssociation_disappears (115.57s)
--- PASS: TestAccAWSEIPAssociation_ec2Classic (246.68s)
--- PASS: TestAccAWSEIPAssociation_instance (109.89s)
--- PASS: TestAccAWSEIPAssociation_networkInterface (73.05s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (117.94s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom (0.00s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom_with_PublicIpv4Pool (0.00s)
--- SKIP: TestAccAWSEIP_CustomerOwnedIPv4Pool (1.45s)
--- SKIP: TestAccAWSEIP_PublicIPv4Pool_custom (0.00s)
```

Output of acceptance tests (`us-east-1`):
```
--- PASS: TestAccAWSEIP_basic (8.96s)
--- PASS: TestAccAWSEIP_BYOIPAddress_default (12.36s)
--- PASS: TestAccAWSEIP_carrierIP (10.35s)
--- PASS: TestAccAWSEIP_disappears (6.82s)
--- PASS: TestAccAWSEIP_Instance (352.90s)
--- PASS: TestAccAWSEIP_Instance_associatedUserPrivateIP (196.47s)
--- PASS: TestAccAWSEIP_Instance_ec2Classic (218.39s)
--- PASS: TestAccAWSEIP_Instance_notAssociated (107.95s)
--- PASS: TestAccAWSEIP_Instance_reassociate (115.58s)
--- PASS: TestAccAWSEIP_networkBorderGroup (9.04s)
--- PASS: TestAccAWSEIP_NetworkInterface (57.57s)
--- PASS: TestAccAWSEIP_NetworkInterface_twoEIPsOneInterface (58.20s)
--- PASS: TestAccAWSEIP_PublicIPv4Pool_default (9.92s)
--- PASS: TestAccAWSEIP_Tags_EC2Classic_withoutVPCTrue (2.92s)
--- PASS: TestAccAWSEIP_Tags_EC2Classic_withVPCTrue (16.32s)
--- PASS: TestAccAWSEIPAssociation_disappears (208.39s)
--- PASS: TestAccAWSEIPAssociation_ec2Classic (218.31s)
--- PASS: TestAccAWSEIPAssociation_instance (302.58s)
--- PASS: TestAccAWSEIPAssociation_networkInterface (55.67s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (301.23s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom (0.00s)
--- SKIP: TestAccAWSEIP_BYOIPAddress_custom_with_PublicIpv4Pool (0.00s)
--- SKIP: TestAccAWSEIP_CustomerOwnedIPv4Pool (0.32s)
--- SKIP: TestAccAWSEIP_PublicIPv4Pool_custom (0.00s)
--- SKIP: TestAccAWSEIP_Tags_EC2VPC_withoutVPCTrue (0.00s)
--- SKIP: TestAccAWSEIP_Tags_EC2VPC_withVPCTrue (0.00s)
--- SKIP: TestAccAWSEIPAssociation_basic (0.00s)
```

**Pull request merged**
